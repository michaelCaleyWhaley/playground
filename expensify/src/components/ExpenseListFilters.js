import React, { Component } from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
} from "../actions/filters";

export class ExpenseListFilters extends Component {
  constructor() {
    super();
    this.state = {
      calendarFocused: null,
    };
  }

  onDatesChange = ({ startDate, endDate }) => {
    const { setStartDate, setEndDate } = this.props;
    setStartDate(startDate);
    setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    this.setState({ calendarFocused });
  };

  onTextChange = (e) => {
    const { setTextFilter } = this.props;
    setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    const { sortByDate, sortByAmount } = this.props;
    if (e.target.value === "date") {
      sortByDate();
    } else if (e.target.value === "amount") {
      sortByAmount();
    }
  };

  render() {
    const { text, sortBy, startDate, endDate } = this.props.filters;
    return (
      <div>
        <input type="text" value={text} onChange={this.onTextChange} />
        <select value={sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDate={startDate}
          startDateId="expensesStartDate"
          endDate={endDate}
          endDateId="expensesEndDate"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (textValue) => dispatch(setTextFilter(textValue)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: () => (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
