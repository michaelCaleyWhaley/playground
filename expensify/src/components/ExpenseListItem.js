import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ description, amount, createdAt, note, id }) => (
  <li>
    <Link to={`/edit/${id}`}>
      <h3>Description: {description}</h3>
    </Link>
    <p>Cost: Â£{amount}</p>
    <p>CreatedAt: {createdAt}</p>
    <p>Note: {note}</p>
  </li>
);

export default ExpenseListItem;
