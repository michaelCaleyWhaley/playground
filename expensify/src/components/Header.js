import React from "react";
import { NavLink } from "react-router-dom";

const Header = () => (
  <header>
    <h1>Expensify App</h1>
    <ul>
      <li>
        <NavLink activeClassName="is-active" exact={true} to="/">
          Dashboard
        </NavLink>
      </li>
      <li>
        <NavLink activeClassName="is-active" exact={true} to="/create">
          Create Expense
        </NavLink>
      </li>
      <li>
        <NavLink activeClassName="is-active" exact={true} to="/help">
          Help
        </NavLink>
      </li>
    </ul>
  </header>
);

export default Header;
