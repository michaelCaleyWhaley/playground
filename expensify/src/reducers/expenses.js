// expenses reducer
// default state
const expensesReducerDefaultState = [];
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(expense => expense.id !== action.id);
      return state;
    case "EDIT_EXPENSE":
      return state.map(currentValue => {
        if (currentValue.id === action.id) {
          return {
            ...currentValue,
            ...action.updates,
          };
        } else {
          return currentValue;
        }
      });
    default:
      return state;
  }
};

export default expensesReducer;
