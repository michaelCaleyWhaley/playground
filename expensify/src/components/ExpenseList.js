import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import SelectExpenses from "../selectors/expenses";

export const ExpenseList = ({ expenses, filters }) => (
  <div>
    <h1>Expense List</h1>
    {filters.text && <p>Filter: {filters.text}</p>}
    <ul>
      {expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        expenses.map((expense, index) => (
          <ExpenseListItem key={index} {...expense} />
        ))
      )}
    </ul>
  </div>
);

const mapStateToProps = (state) => ({
  ...state,
  expenses: SelectExpenses(state.expenses, state.filters),
});
export default connect(mapStateToProps)(ExpenseList);
