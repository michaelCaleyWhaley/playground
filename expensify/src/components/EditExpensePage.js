import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    const { props } = this;
    props.editExpense(props.expense.id, expense);
    props.history.push("/");
  };

  onClick = () => {
    const { props } = this;
    props.removeExpense(props.expense.id);
    props.history.push("/");
  };

  render() {
    const { props } = this;

    return (
      <div>
        <ExpenseForm expense={props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  removeExpense: (expenseId) => {
    dispatch(removeExpense({ id: expenseId }));
  },
  editExpense: (expenseId, expense) => {
    dispatch(editExpense(expenseId, expense));
  },
});

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === props.match.params.id
    ),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
