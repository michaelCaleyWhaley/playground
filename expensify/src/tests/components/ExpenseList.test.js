import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("should render ExpenseList with expenses", () => {
  const wrapper = shallow(
    <ExpenseList expenses={expenses} filters={{ text: "test" }} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseList with empty message", () => {
  const wrapper = shallow(
    <ExpenseList expenses={[]} filters={{ text: "test" }} />
  );
  expect(wrapper).toMatchSnapshot();
});
