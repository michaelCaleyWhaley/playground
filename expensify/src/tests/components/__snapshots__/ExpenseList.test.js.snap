// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    Filter: 
    test
  </p>
  <ul>
    <p>
      No expenses
    </p>
  </ul>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    Filter: 
    test
  </p>
  <ul>
    <ExpenseListItem
      amount={195}
      createdAt={"1970-01-01T00:00:00.000Z"}
      description="Gum"
      id="1"
      key="0"
      note=""
    />
    <ExpenseListItem
      amount={1095}
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={4500}
      createdAt={864000000}
      description="Credit card"
      id="3"
      key="2"
      note=""
    />
    <ExpenseListItem
      amount={200}
      createdAt={-172800000}
      description="Monitor"
      id="4"
      key="3"
      note=""
    />
  </ul>
</div>
`;
